src/Jaeger.Core/Senders/ISender.cs

```diff
@@ -1,11 +1,15 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
 namespace Jaeger.Core.Senders
 {
     public interface ISender
     {
-        int Append(Span span);
+        Task<int> AppendAsync(Span span, CancellationToken cancellationToken);
```

Comment:
We should use `CancellationToken cancellationToken = default` as parameter. According to MSDN, `default(CancellationToken)` is semantically equivalent to `CancellationToken.None` and the source code shows, that  `CancellationToken.None` just does exactly return `default(CancellationToken)`. With this, the implementors of the interface don't have to offer both variants and `AppendCommand` and `FlushCommand` in `RemoteReporter` don't have to use `CancellationToken.None` directly, which always looks like a workaround.

Reply:
I disagree on this one. While optional parameters are quite nice on concrete classes, they are confusing as hell on interfaces - see e.g. [link 1](https://blogs.msdn.microsoft.com/ericlippert/2011/05/09/optional-argument-corner-cases-part-one/), [link 2](http://blackrabbitcoder.net/archive/2010/06/17/c-optional-parameters---pros-and-pitfalls.aspx), [link 3](https://stackoverflow.com/questions/4922714/why-are-c-sharp-4-optional-parameters-defined-on-interface-not-enforced-on-imple#4923642). There's even a code analysis rule that warns when optional parameters are used on public types: [CA1026](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1026-default-parameters-should-not-be-used).

The thing with making the cancellation token optional is that as soon as that happens, no one uses it anymore. Keeping it as required forces people to either supply it or to make the conscious  decision to ignore it (by passing `CancellationToken.None`). This also makes it much easier to search for such cases because you can always search for "CancellationToken.None".

We should rather look at my current usages of `CancellationToken.None` and check if we should pass some timeout instead.